File: packages/yoke-core/src/yoke_core/ui/static/session_usage_display.js
/**
 * Show what a session consumed without implying a precision it lacks.
 *
 * The control plane ships derived figures rather than the stored reading,
 * so pricing happens once server-side and a card, a table and a summed
 * tile cannot disagree. What is left here is presentation, and its whole
 * job is distinguishing three states an operator would otherwise conflate:
 * a measured figure, a figure computed from an incomplete reading, and no
 * measurement at all. The last is never drawn as a zero — a session that
 * has not been read yet did not run for free.
 *
 * The dollar figure is an estimate of API-equivalent spend. A session run
 * under a subscription plan spends that plan's own meters instead, and no
 * published conversion turns one into the other, so the label says so
 * everywhere the number appears.
 */

// Marks a figure computed from an incomplete reading. Trails the value so a
// column of numbers still lines up on its digits.
export const PARTIAL_MARK = "~";

// What a session with no reading at all shows.
export const UNREAD_DISPLAY = "—";

const COMPLETE = "complete";

function finiteCount(value) {
  if (value === null || value === undefined || value === "") return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : null;
}

export function compactTokens(value) {
  const count = finiteCount(value);
  if (count === null) return "";
  const scale = count >= 1_000_000
    ? [1_000_000, "m"]
    : count >= 1_000 ? [1_000, "k"] : null;
  if (!scale) return String(Math.round(count));
  const scaled = count / scale[0];
  const precision = scaled < 10 ? 1 : 0;
  return `${scaled.toFixed(precision).replace(/\.0$/, "")}${scale[1]}`;
}

export function compactUsd(value) {
  const amount = finiteCount(value);
  if (amount === null) return "";
  if (amount < 1) return `$${amount.toFixed(2)}`;
  if (amount < 100) return `$${amount.toFixed(2)}`.replace(/\.?0+$/, "");
  return `$${Math.round(amount).toLocaleString("en-US")}`;
}

function mark(status) {
  return status && status !== COMPLETE ? PARTIAL_MARK : "";
}

export function sessionTokensDisplay(row) {
  const tokens = compactTokens(row?.usage_tokens);
  return tokens ? `${tokens}${mark(row?.usage_status)}` : UNREAD_DISPLAY;
}

export function sessionCostDisplay(row) {
  const cost = compactUsd(row?.usage_cost_usd);
  return cost ? `${cost}${mark(row?.usage_cost_status)}` : UNREAD_DISPLAY;
}

export function sessionUsageIsPartial(row) {
  return Boolean(
    mark(row?.usage_status) || (row?.usage_cost_usd && mark(row?.usage_cost_status)),
  );
}

/**
 * Sum the sessions a tile is actually showing, and say how many that was.
 *
 * A machine tile answers for the rows currently on the page, not for the
 * machine's whole history, so the scope travels with the total: `covered`
 * counts the rows that carried a reading and `total` counts the rows the
 * sum ranged over. `covered` is what the tile's SESSIONS count names.
 */
export function summarizeSessionUsage(rows) {
  const scoped = Array.isArray(rows) ? rows : [];
  let tokens = 0;
  let cost = 0;
  let covered = 0;
  let costed = 0;
  let partial = false;
  for (const row of scoped) {
    const rowTokens = finiteCount(row?.usage_tokens);
    const rowCost = finiteCount(row?.usage_cost_usd);
    if (rowTokens !== null) {
      tokens += rowTokens;
      covered += 1;
      if (mark(row?.usage_status)) partial = true;
    }
    if (rowCost !== null) {
      cost += rowCost;
      costed += 1;
      if (mark(row?.usage_cost_status)) partial = true;
    }
  }
  // A sum missing some of its members is itself partial, whatever each
  // member's own reading said.
  if (covered < scoped.length || costed < covered) partial = true;
  return {
    tokens,
    cost,
    covered,
    costed,
    total: scoped.length,
    partial: covered > 0 && partial,
  };
}

export function usageSummaryLabel(summary) {
  if (!summary || summary.covered === 0) {
    return summary?.total
      ? `no consumption recorded for ${summary.total} ` +
        `${summary.total === 1 ? "session" : "sessions"}`
      : "";
  }
  const suffix = summary.partial ? PARTIAL_MARK : "";
  const money = summary.costed > 0
    ? ` · ${compactUsd(summary.cost)}${suffix}`
    : "";
  return `${compactTokens(summary.tokens)}${suffix}${money}`;
}

// A machine tile is its own rollup, not one session's reading, so it
// states a total rather than a hedge — the SESSIONS count beside it
// already says how much of the roster that total answers for.
export function usageSummaryTokenDisplay(summary) {
  if (!summary || summary.covered === 0) return UNREAD_DISPLAY;
  return compactTokens(summary.tokens);
}

export function usageSummaryCostDisplay(summary) {
  if (!summary || summary.costed === 0) return UNREAD_DISPLAY;
  return compactUsd(summary.cost);
}

/**
 * Name the priced count only when it trails the session count already
 * shown on the tile.
 *
 * Token coverage and cost coverage are different counts and routinely
 * differ: a session on a model nobody has researched a price for reports
 * its tokens and contributes nothing to the dollar figure. Repeating an
 * identical count teaches nothing, so this stays silent until the two
 * diverge.
 */
export function usageSummaryScope(summary) {
  if (!summary || !summary.costed || summary.costed === summary.covered) return "";
  return `${summary.costed} priced`;
}

File: packages/yoke-contracts/src/yoke_contracts/session_usage_display.py
"""One rendering of session consumption, shared by every surface.

The board, the session roster, and the workbench all answer the same two
questions — how many tokens, and what would they have cost — so they
answer them with the same words. A figure that reads ``1.2m`` on the
board and ``1,234,567`` in the roster invites the reader to wonder
whether they are the same measurement.

Every rendering here is honest about what it does not know. An unread
session shows nothing rather than a zero; a partial reading is marked;
and a cost that could not be priced shows why rather than a dash that
could equally mean "free". The dollar figure always reads as an estimate
of API-equivalent spend, never as plan consumption.
"""

from __future__ import annotations

from typing import Optional

from yoke_contracts.session_usage_cost import COST_COMPLETE, SessionCost
from yoke_contracts.session_usage_facts import USAGE_COMPLETE, SessionUsage


#: Marks a figure computed from an incomplete reading. Placed after the
#: value so a column of numbers still lines up on the digits.
PARTIAL_MARK = "~"

#: What a session with no reading at all shows.
UNREAD_DISPLAY = "—"


def compact_tokens(count: int) -> str:
    """Render a token count at the scale an operator reads it in."""
    if count <= 0:
        return "0"
    if count >= 1_000_000:
        scaled, suffix = count / 1_000_000, "m"
    elif count >= 1_000:
        scaled, suffix = count / 1_000, "k"
    else:
        return str(count)
    precision = 1 if scaled < 10 else 0
    return f"{scaled:.{precision}f}".rstrip("0").rstrip(".") + suffix


def compact_usd(amount: float) -> str:
    """Render an estimate at a precision the estimate actually supports."""
    if amount <= 0:
        return "$0"
    if amount < 1:
        return f"${amount:.2f}"
    if amount < 100:
        return f"${amount:.2f}".rstrip("0").rstrip(".")
    return f"${amount:,.0f}"


def tokens_display(usage: Optional[SessionUsage]) -> str:
    """Render one session's token total, marked when partly unknown."""
    if usage is None or not usage.models:
        return UNREAD_DISPLAY
    total = usage.billable_tokens()
    if total <= 0:
        return UNREAD_DISPLAY
    mark = "" if usage.status == USAGE_COMPLETE else PARTIAL_MARK
    return f"{compact_tokens(total)}{mark}"


def cost_display(cost: Optional[SessionCost]) -> str:
    """Render one session's estimate, marked when partly unpriced."""
    if cost is None or not cost.priced():
        return UNREAD_DISPLAY
    mark = "" if cost.status == COST_COMPLETE else PARTIAL_MARK
    return f"{compact_usd(cost.usd)}{mark}"


def usage_cell(usage: Optional[SessionUsage], cost: Optional[SessionCost]) -> str:
    """Render both figures as one compact cell for a table."""
    tokens = tokens_display(usage)
    money = cost_display(cost)
    if tokens == UNREAD_DISPLAY and money == UNREAD_DISPLAY:
        return UNREAD_DISPLAY
    return f"{tokens} · {money}"


def usage_title(usage: Optional[SessionUsage], cost: Optional[SessionCost]) -> str:
    """The full explanation behind a compact cell, for a tooltip or note.

    Everything a compact figure had to drop lives here: which reading was
    partial and why, where the prices came from and when they were
    checked, and the standing caveat that this is API-equivalent spend
    rather than consumption of any subscription plan.
    """
    parts: list[str] = []
    if usage is None or not usage.models:
        parts.append("no consumption recorded for this session yet")
    else:
        parts.append(f"{usage.billable_tokens():,} tokens recorded")
        if usage.source:
            parts.append(f"read from {usage.source}")
        if usage.status != USAGE_COMPLETE and usage.reason:
            parts.append(f"partial: {usage.reason}")
        if usage.cost_caveat:
            parts.append(f"tokens exact, pricing uncertain: {usage.cost_caveat}")
    if cost is not None and cost.priced():
        parts.append("estimated API-equivalent cost, not plan consumption")
        if cost.price_basis:
            parts.append(f"prices from {cost.price_basis}")
        if cost.effective_date:
            parts.append(f"effective {cost.effective_date}")
        if cost.checked_date:
            parts.append(f"checked {cost.checked_date}")
        if cost.status != COST_COMPLETE and cost.reason:
            parts.append(f"partial: {cost.reason}")
    elif cost is not None and cost.reason:
        parts.append(f"no cost estimate: {cost.reason}")
    return " · ".join(parts)


def usage_projection(usage_totals: object) -> dict[str, object]:
    """Render-ready consumption facts for one stored reading.

    Surfaces receive derived values rather than the stored document, so
    the pricing rules live in one place and a table, a card, and a summed
    tile cannot disagree about what a session cost. ``None`` in either
    figure means "not measured" and is what a caller must not render as a
    zero; ``note`` always explains whatever the compact figures dropped.
    """
    from yoke_contracts.session_usage_facts import usage_from_document
    from yoke_contracts.session_usage_pricing import estimated_session_cost

    usage = usage_from_document(usage_totals)
    cost = estimated_session_cost(usage)
    measured = usage is not None and usage.billable_tokens() > 0
    return {
        "usage_tokens": usage.billable_tokens() if measured else None,
        "usage_status": usage.status if usage is not None else None,
        "usage_cost_usd": cost.usd if cost.priced() else None,
        "usage_cost_status": cost.status,
        "usage_note": usage_title(usage, cost),
    }


#: The keys :func:`usage_projection` returns, for read-model field lists.
USAGE_PROJECTION_FIELDS = (
    "usage_tokens",
    "usage_status",
    "usage_cost_usd",
    "usage_cost_status",
    "usage_note",
)


__all__ = [
    "PARTIAL_MARK",
    "USAGE_PROJECTION_FIELDS",
    "UNREAD_DISPLAY",
    "compact_tokens",
    "compact_usd",
    "cost_display",
    "tokens_display",
    "usage_cell",
    "usage_projection",
    "usage_title",
]
